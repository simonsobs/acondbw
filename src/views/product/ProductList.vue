<template>
  <v-container class="product-list" style="position: relative">
    <v-row>
      <v-col class="pb-0">
        <v-card-actions class="align-end" style="height: 56px">
          <v-tooltip bottom open-delay="800">
            <template v-slot:activator="{ on }">
              <v-btn :disabled="loading" icon @click="refresh()" v-on="on">
                <v-icon>mdi-refresh</v-icon>
              </v-btn>
            </template>
            <span>Refresh</span>
          </v-tooltip>
          <v-spacer></v-spacer>
          <span v-if="loaded" class="secondary--text">
            <span v-if="nItemsTotal == 1">1 {{ productType.singular }}</span>
            <span v-else>{{ nItemsTotal }} {{ productType.plural }}</span>
          </span>
          <v-spacer></v-spacer>
          <span v-if="loaded && nItemsTotal > 1">
            <v-tooltip bottom open-delay="800">
              <template v-slot:activator="{ on: tooltip }">
                <v-menu left offset-y>
                  <template v-slot:activator="{ on: menu }">
                    <v-btn icon v-on="{ ...tooltip, ...menu }">
                      <v-icon>mdi-sort-variant</v-icon>
                    </v-btn>
                  </template>
                  <v-list flat dense>
                    <v-list-item>
                      <v-list-item-title>Sort</v-list-item-title>
                    </v-list-item>
                    <v-list-item-group v-model="sortItem" color="primary">
                      <v-list-item v-for="(item, i) in sortItems" :key="i">
                        <v-list-item-icon>
                          <v-icon v-if="i == sortItem">mdi-check</v-icon>
                        </v-list-item-icon>
                        <v-list-item-content>
                          <v-list-item-title
                            v-text="item.text"
                          ></v-list-item-title>
                        </v-list-item-content>
                      </v-list-item>
                    </v-list-item-group>
                  </v-list>
                </v-menu>
              </template>
              <span>Sort</span>
            </v-tooltip>
          </span>
          <v-spacer></v-spacer>
          <span v-if="loaded">
            <v-tooltip bottom open-delay="800">
              <template v-slot:activator="{ on }">
                <v-btn
                  icon
                  @click="areAllCardsCollapsed = !areAllCardsCollapsed"
                  v-on="on"
                >
                  <v-icon>
                    {{
                      areAllCardsCollapsed
                        ? "mdi-unfold-more-horizontal"
                        : "mdi-unfold-less-horizontal"
                    }}
                  </v-icon>
                </v-btn>
              </template>
              <span>
                {{ areAllCardsCollapsed ? "Expand all" : "Collapse all" }}
              </span>
            </v-tooltip>
          </span>
          <span v-if="loaded || empty">
            <v-btn
              :disabled="disableAdd"
              fab
              class="ml-3 secondary"
              :to="{
                name: 'ProductAdd',
                params: { productTypeName: productType.name },
              }"
            >
              <v-icon class="on-secondary--text">mdi-plus-thick</v-icon>
            </v-btn>
          </span>
        </v-card-actions>
      </v-col>
    </v-row>
    <v-row>
      <v-col v-if="loading">
        <v-progress-circular
          indeterminate
          :size="26"
          color="secondary"
        ></v-progress-circular>
      </v-col>
      <v-col v-if="loaded" class="pt-0 pb-16">
        <component
          :is="productItemCard"
          v-for="edge in edges"
          :key="edge.node.id"
          :productId="edge.node.productId"
          collapsible="true"
          :collapsed="isCardCollapsed[edge.node.id]"
          v-on:expand="isCardCollapsed[edge.node.id] = false"
          v-on:collapse="isCardCollapsed[edge.node.id] = true"
          :disableEdit="disableEdit"
          :disableDelete="disableDelete"
          class="my-1"
        ></component>
        <div v-if="loadingMore" class="pa-3">
          <v-progress-circular
            indeterminate
            :size="26"
            color="secondary"
          ></v-progress-circular>
        </div>
        <v-card-actions v-if="showLoadMoreButton">
          <v-spacer></v-spacer>
          <span v-if="nItemsTotal > 1">
            <span v-if="edges.length == nItemsTotal">
              {{ nItemsTotal }} {{ productType.plural }}
            </span>
            <span v-else>
              {{ edges.length }} of {{ nItemsTotal }}
              {{ productType.plural }}
            </span>
          </span>
          <v-spacer></v-spacer>
          <v-btn
            :disabled="!productType.products.pageInfo.hasNextPage"
            outlined
            color="secondary"
            text
            @click="loadMore()"
          >
            Load more
          </v-btn>
        </v-card-actions>
      </v-col>
      <v-col v-else-if="empty">
        <v-card-text>
          Empty. No {{ productType.plural }} are found.
        </v-card-text>
      </v-col>
      <v-col v-else-if="error">
        <v-alert v-if="error" type="error">{{ error }}</v-alert>
      </v-col>
      <v-col v-else>
        <!-- state = State.NONE -->
      </v-col>
    </v-row>
    <dev-tool-loading-state-overriding-menu
      @state="devtoolState = $event"
    ></dev-tool-loading-state-overriding-menu>
  </v-container>
</template>

<script>
import ProductItemCard from "@/components/product/ProductItemCard.vue";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/utils/DevToolLoadingStateOverridingMenu.vue";

import QueryForProductList from "@/graphql/queries/QueryForProductList.gql";

export default {
  name: "ProductList",
  components: {
    ProductItemCard,
    DevToolLoadingStateOverridingMenu,
  },
  props: {
    productTypeId: { required: true },
    productItemCard: { default: "ProductItemCard" },
    productAddForm: { default: "ProductAddForm" },
    disableAdd: { default: false },
    disableEdit: { default: false },
    disableDelete: { default: false },
  },
  data() {
    return {
      productType: null,
      init: true,
      error: null,
      refreshing: false,
      loadingMore: false,
      sortItem: 0,
      sortItems: [
        { text: "Recently produced", value: "DATE_PRODUCED_DESC" },
        { text: "Recently posted", value: "DATE_POSTED_DESC" },
        { text: "Recently updated", value: "DATE_UPDATED_DESC" },
        { text: "Name", value: "NAME_ASC" },
      ],
      nItemsInitialLoad: 10,
      nEtraItemsAutomaticLoad: 2,
      nItemsPerLoad: 20,
      isCardCollapsed: {},
      devtoolState: null,
      State: State,
    };
  },
  apollo: {
    productType: {
      query: QueryForProductList,
      variables() {
        return {
          typeId: this.productTypeId,
          sort: [this.sort],
          first: this.nItemsInitialLoad,
          after: "",
        };
      },
      skip: function () {
        return !this.productTypeId || !this.sort;
      },
      result(result) {
        this.init = false;
        this.error = result.error ? result.error : null;
      },
    },
  },
  computed: {
    edges() {
      return this.productType ? this.productType.products.edges : null;
    },
    nItemsTotal() {
      return this.productType ? this.productType.products.totalCount : null;
    },
    showLoadMoreButton() {
      if (this.loadingMore) {
        return false;
      }
      return (
        this.nItemsTotal > this.nItemsInitialLoad + this.nEtraItemsAutomaticLoad
      );
    },
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.refreshing) {
        return State.LOADING;
      }

      if (this.edges) {
        if (this.edges.length) {
          return State.LOADED;
        } else {
          return State.EMPTY;
        }
      } else if (this.$apollo.queries.productType.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.state == State.LOADING;
    },
    loaded() {
      return this.state == State.LOADED;
    },
    empty() {
      return this.state == State.EMPTY;
    },
    notFound() {
      return this.state == State.NONE;
    },
    sort() {
      return this.sortItems[this.sortItem].value;
    },
    areAllCardsCollapsed: {
      get: function () {
        return Object.keys(this.isCardCollapsed).every(
          (i) => this.isCardCollapsed[i]
        );
      },
      set: function (v) {
        for (const k in this.isCardCollapsed) {
          this.isCardCollapsed[k] = v;
        }
      },
    },
  },
  watch: {
    devtoolState: function () {
      if (this.devtoolState) {
        this.init = this.devtoolState == State.INIT;
      }
      this.error =
        this.devtoolState == State.ERROR ? "Error from Dev Tools" : null;
    },
    "$store.state.nApolloMutations": function () {
      this.refresh();
    },
    edges: function () {
      this.collapseCards();
      this.loadAllFewRemainingItems();
    },
  },
  methods: {
    collapseCards() {
      if (this.edges == undefined) {
        return;
      }
      for (const edge of this.edges) {
        const id = edge.node.id;
        if (!(id in this.isCardCollapsed)) {
          this.isCardCollapsed = {
            ...this.isCardCollapsed,
            [id]: true,
          };
          // The above line of the code adds a new element {id: true} to
          // the object this.isCardCollapsed in the way that the new
          // element will be a reactive object of Vue. The commented out
          // code below is simpler but the new element won't be reactive.
          // this.isCardCollapsed[id] = true;
        }
      }
    },
    loadAllFewRemainingItems() {
      if (this.edges == undefined) {
        return;
      }
      if (
        this.edges.length + this.nEtraItemsAutomaticLoad >=
        this.nItemsTotal
      ) {
        this.loadMore();
      }
    },
    async refresh() {
      this.refreshing = true;
      const wait = new Promise((resolve) => setTimeout(resolve, 500));
      await this.$apollo.queries.productType.refetch();
      this.areAllCardsCollapsed = true;
      await wait; // wait until 0.5 sec passes since starting refetch
      // because the progress circular is too flickering if
      // the refetch finishes too quickly
      this.refreshing = false;
    },
    async loadMore() {
      if (!this.productType.products.pageInfo.hasNextPage) {
        return;
      }

      this.loadingMore = true;
      try {
        await this.$apollo.queries.productType.fetchMore({
          variables: {
            first: this.nItemsPerLoad,
            after: this.productType.products.pageInfo.endCursor,
          },
          updateQuery: (previousResult, { fetchMoreResult }) => {
            return {
              productType: {
                ...previousResult.productType,
                products: {
                  ...previousResult.productType.products,
                  pageInfo: {
                    ...previousResult.productType.products.pageInfo,
                    endCursor:
                      fetchMoreResult.productType.products.pageInfo.endCursor,
                    hasNextPage:
                      fetchMoreResult.productType.products.pageInfo.hasNextPage,
                  },
                  edges: [
                    ...previousResult.productType.products.edges,
                    ...fetchMoreResult.productType.products.edges,
                  ],
                },
              },
            };
          },
        });
      } catch (error) {}
      this.loadingMore = false;
    },
  },
};
</script>
