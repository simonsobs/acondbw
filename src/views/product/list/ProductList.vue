<template>
  <div class="my-5" style="block-size: 100%; position: relative">
    <div v-if="error">
      <v-alert variant="tonal" type="error" :text="error"> </v-alert>
    </div>
    <top-bar
      :loading="loading"
      :loaded="loaded"
      :product-type="productType"
      v-model:sort="sort"
      v-model:are-all-cards-collapsed="areAllCardsCollapsed"
      @refresh="refresh()"
    >
    </top-bar>
    <div>
      <div v-if="nodes.length" class="pt-0 pb-16">
        <list
          :loading="loading"
          :product-type="productType"
          :edges="edges"
          :nodes="nodes"
          :n-items-total="nItemsTotal"
          :n-items-initial-load="nItemsInitialLoad"
          v-model:first="first"
          :is-card-collapsed="isCardCollapsed"
          v-if="!empty && productType"
        >
        </list>
      </div>
      <div v-else-if="loading">
        <div class="pa-3">
          <v-progress-circular
            indeterminate
            :size="26"
            color="secondary"
          ></v-progress-circular>
        </div>
      </div>
      <div v-else-if="empty">
        <empty :product-type="productType" v-if="productType"></empty>
      </div>
    </div>
    <v-btn
      :disabled="disableAdd"
      icon
      variant="flat"
      size="x-large"
      color="tertiary-container"
      elevation="8"
      :to="{
        name: 'ProductAdd',
        params: { productTypeName: productType.name },
      }"
      v-if="loaded && !empty && productType"
      style="position: fixed; bottom: 24px; right: 24px"
    >
      <v-icon icon="mdi-plus-thick"></v-icon>
    </v-btn>
    <dev-tool-loading-state-menu top="0" right="-35px" v-model="devtoolState">
    </dev-tool-loading-state-menu>
  </div>
</template>

<script setup lang="ts">
import { ref, reactive, watch, computed } from "vue";
import { storeToRefs } from "pinia";
import { useStore } from "@/plugins/pinia/stores/main";
import { useConfigStore } from "@/plugins/pinia/stores/config";

import TopBar from "./TopBar.vue";
import List from "./List.vue";
import Empty from "./Empty.vue";

import {
  useQueryForProductListQuery,
  ProductSortEnum,
} from "@/graphql/codegen/generated";

import { useQueryState } from "@/utils/query-state";

interface Props {
  productTypeId: number;
}

const props = defineProps<Props>();

const configStore = useConfigStore();
const disableAdd = computed(() => !configStore.config.productCreationDialog);

const nItemsInitialLoad = ref(10);
const first = ref(nItemsInitialLoad.value);

const sort = ref<ProductSortEnum[]>([]);

const query = useQueryForProductListQuery({
  variables: {
    typeId: props.productTypeId,
    // @ts-ignore
    sort: sort,
    // @ts-ignore
    first: first,
    after: "",
  },
});

type Query = typeof query;

const productType = computed(() => query.data?.value?.productType);

function readEdges(query: Query) {
  const edgesAndNulls = query.data?.value?.productType?.products?.edges;
  if (!edgesAndNulls) return [];
  return edgesAndNulls.flatMap((e) => (e ? [e] : []));
}
function readNodes(query: Query) {
  return readEdges(query).flatMap((e) => (e.node ? e.node : []));
}

function isEmpty(query: Query) {
  return readNodes(query).length === 0;
}

const queryState = useQueryState(query, { isEmpty });

const {
  loading,
  loaded,
  error,
  empty,
  devtoolState,
  refresh: refresh_,
} = queryState;

const refreshing = ref(false);

const edges = computed(() =>
  refreshing.value || empty.value ? [] : readEdges(query)
);

const nodes = computed(() =>
  refreshing.value || empty.value ? [] : readNodes(query)
);

const nItemsTotal = computed(
  () => productType.value?.products?.totalCount || 0
);

async function refresh() {
  refreshing.value = true;
  first.value = nItemsInitialLoad.value;
  try {
    await refresh_();
  } finally {
    areAllCardsCollapsed.value = true;
    refreshing.value = false;
  }
}

const isCardCollapsed = reactive<{ [key: string]: boolean }>({});

const areAllCardsCollapsed = computed({
  get: () => Object.values(isCardCollapsed).every((i) => i),
  set: (v) => {
    Object.keys(isCardCollapsed).forEach((k) => {
      isCardCollapsed[k] = v;
    });
  },
});

const store = useStore();
const { nApolloMutations } = storeToRefs(store);
watch(nApolloMutations, refresh);
</script>
