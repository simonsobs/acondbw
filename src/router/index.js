import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
