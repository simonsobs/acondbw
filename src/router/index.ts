import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";
import VueMeta from "vue-meta";

import { default as defaultPinia } from "@/stores";
import { useAuthStore } from "@/stores/auth";

import Frame from "@/components/layout/Frame.vue";
import NullFrame from "@/components/layout/NullFrame.vue";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";
import SearchTest from "@/views/framework/SearchTest.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue";

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import ProductType from "@/views/admin/ProductType.vue";

const About = () => import("@/views/framework/About.vue");
const NotFound = () => import("@/views/framework/NotFound.vue");
const AccessDenied = () => import("@/views/framework/AccessDenied.vue");

const FrameAdmin = () =>
  import(/* webpackChunkName: "admin" */ "@/components/admin/FrameAdmin.vue");
const Version = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Version.vue");
const Log = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Log.vue");
const Config = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Config.vue");
const Theme = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Theme.vue");
const User = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/User.vue");
const AdminAppToken = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken.vue"
  );
const AdminAppAuth = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth.vue"
  );
const AdminAppTokenError = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError.vue"
  );

Vue.use(VueRouter);
Vue.use(VueMeta);

let pinia = defaultPinia;

function setPinia(val) {
  // because not sure how to monkey patch pinia for tests
  pinia = val;
}

function setDefaultPinia() {
  setPinia(defaultPinia);
}

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    // meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/search-test",
    name: "SearchTest",
    components: { default: SearchTest, frame: Frame },
    meta: { requiresAuth: true },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: { default: OAuthRedirect, frame: NullFrame },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: { default: SignIn, frame: NullFrame },
  },
  {
    path: "/auth",
    name: "Auth",
    components: { default: Auth, frame: NullFrame },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: { default: SignInError, frame: NullFrame },
  },
  {
    path: "/admin",
    name: "Admin",
    redirect: { name: "AdminVersion" },
  },
  {
    path: "/admin/product-type",
    name: "ProductType",
    components: { default: ProductType, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/versions",
    name: "AdminVersion",
    components: { default: Version, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/log",
    name: "AdminLog",
    components: { default: Log, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/config",
    name: "AdminConfig",
    components: { default: Config, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/theme",
    name: "AdminTheme",
    components: { default: Theme, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/users",
    name: "AdminUser",
    components: { default: User, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: { default: AdminAppToken, frame: FrameAdmin },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: { default: AdminAppAuth, frame: FrameAdmin },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: { default: AdminAppTokenError, frame: FrameAdmin },
  },
  {
    path: "/about",
    name: "About",
    component: About,
  },
  {
    path: "/add-product/:productTypeName",
    name: "ProductAdd",
    components: { default: ProductAdd, frame: NullFrame },
    meta: { requiresAuth: true },
  },
  {
    path: "/product/:productTypeName",
    components: { default: ProductTop, frame: Frame },
    meta: { requiresAuth: true },
    children: [
      { path: "", name: "ProductList", component: ProductList },
      { path: ":name", name: "ProductItem", component: ProductItem },
    ],
  },
  {
    path: "/access-denied",
    name: "AccessDenied",
    components: { default: AccessDenied, frame: NullFrame },
  },
  {
    path: "*",
    name: "NotFound",
    components: { default: NotFound, frame: NullFrame },
  },
];

function createRouter() {
  const router = new VueRouter({
    mode: "history",
    base: process.env.BASE_URL,
    routes,
  });

  router.beforeEach((to, from, next) => {
    const auth = useAuthStore(pinia);

    const adminRequired = to.matched.some(
      (record) => record.meta.requiresAdmin
    );
    const isAdmin = auth.isAdmin;
    if (adminRequired && !isAdmin) {
      next({ name: "AccessDenied" });
      return;
    }

    const authRequired = to.matched.some((record) => record.meta.requiresAuth);
    const isSignedIn = auth.isSignedIn as boolean;
    if (authRequired && !isSignedIn) {
      next({ name: "AccessDenied" });
      return;
    }
    next();
  });

  return router;
}

const router = createRouter();

function checkAuthForCurrentRoute() {
  const authRequired = router.currentRoute.matched.some(
    (record) => record.meta.requiresAuth
  );
  const auth = useAuthStore(pinia);
  const signedIn = auth.isSignedIn as boolean;
  if (authRequired && !signedIn) {
    router.push("/");
  }
}

export { createRouter, checkAuthForCurrentRoute, setPinia, setDefaultPinia };
