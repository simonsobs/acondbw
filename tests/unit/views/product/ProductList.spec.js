import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";

import ProductList from "@/views/product/ProductList.vue";

import router from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("ProductList.vue", () => {
  let localVue;

  function createWrapper(loading = false) {
    return mount(ProductList, {
      localVue,
      router,
      mocks: {
        $apollo: {
          queries: {
            productType: {
              loading: loading,
            },
          },
        },
      },
      propsData: {
        productTypeId: 1,
        disableAdd: false,
      },
      stubs: {
        ProductItemCard: true,
        ProductAddForm: true,
        "dev-tool-loading-state-overriding-menu": true,
        "transition-group": true,
      },
    });
  }

  const edges = [
    {
      node: {
        id: "TWFwOjEwMTM=",
        productId: "1013",
        name: "lat20200201",
      },
    },
    {
      node: {
        id: "TWFwOjEwMTI=",
        productId: "1012",
        name: "lat20200120",
      },
    },
    {
      node: {
        id: "TWFwOjEwMDE=",
        productId: "1001",
        name: "lat20190213",
      },
    },
  ];

  const productType = {
    id: "UHJvZHVjdFR5cGU6MQ==",
    typeId: "1",
    name: "map",
    singular: "map",
    plural: "maps",
    indefArticle: "a",
    products: {
      totalCount: 3,
      pageInfo: {
        hasNextPage: true,
      },
      edges: edges,
    },
  };

  beforeEach(function () {
    localVue = createLocalVue();
  });

  it("match snapshot", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      productType: productType,
    });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("loading", async () => {
    const loading = true;
    const wrapper = createWrapper(loading);
    await Vue.nextTick();
    expect(wrapper.find(".v-progress-circular").exists()).toBe(true);
  });

  it("error", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      error: "Error: cannot load data",
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Error: cannot load data");
  });

  it("none", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      productType: { ...productType, ...{ products: { edges: [] } } },
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Empty. No maps are found.");
  });
});
