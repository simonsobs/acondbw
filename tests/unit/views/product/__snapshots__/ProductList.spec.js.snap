// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList.vue match snapshot 1`] = `
<v-container-stub tag="div" class="product-list" style="position: relative;">
  <v-row-stub tag="div">
    <v-col-stub tag="div" class="pb-0">
      <v-card-actions-stub class="align-end" style="height: 56px;">
        <v-tooltip-stub opendelay="800" closedelay="0" bottom="true" openonhover="true" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" tag="span"> <span>Refresh</span></v-tooltip-stub>
        <v-spacer-stub></v-spacer-stub> <span class="secondary--text"><span>3 maps</span></span>
        <v-spacer-stub></v-spacer-stub> <span><v-tooltip-stub opendelay="800" closedelay="0" bottom="true" openonhover="true" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" tag="span"> <span>Sort</span></v-tooltip-stub></span>
        <v-spacer-stub></v-spacer-stub> <span><v-tooltip-stub opendelay="800" closedelay="0" bottom="true" openonhover="true" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" tag="span"> <span>
              Expand all
            </span></v-tooltip-stub></span> <span><v-btn-stub tag="button" activeclass="" to="[object Object]" fab="true" type="button" class="ml-3 secondary"><v-icon-stub class="on-secondary--text">mdi-plus-thick</v-icon-stub></v-btn-stub></span>
      </v-card-actions-stub>
    </v-col-stub>
  </v-row-stub>
  <v-row-stub tag="div">
    <!---->
    <v-col-stub tag="div" class="pt-0 pb-16">
      <productitemcard-stub productid="1013" collapsed="true" collapsible="true" class="my-1"></productitemcard-stub>
      <productitemcard-stub productid="1012" collapsed="true" collapsible="true" class="my-1"></productitemcard-stub>
      <productitemcard-stub productid="1001" collapsed="true" collapsible="true" class="my-1"></productitemcard-stub>
      <!---->
      <!---->
    </v-col-stub>
  </v-row-stub>
  <dev-tool-loading-state-overriding-menu-stub></dev-tool-loading-state-overriding-menu-stub>
</v-container-stub>
`;
